RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+            int rhIdx = pfClusteringVars[pfClusteringVars[topoId].topoSeedOffsets()]
+                            .topoSeedList();  // i is the seed rechit index
+            int seedIdx = pfClusteringVars[rhIdx].rhIdxToSeedIdx();
+            clusterView[seedIdx].energy() = pfRecHits[rhIdx].energy();
+            clusterView[seedIdx].x() = pfRecHits[rhIdx].x();
+            clusterView[seedIdx].y() = pfRecHits[rhIdx].y();
+            clusterView[seedIdx].z() = pfRecHits[rhIdx].z();
+          }
+        } else if (nSeeds == 1) {
+          // Single seed cluster
+          hcalFastCluster_singleSeed(
+              acc, pfClusParams, topoId, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 100 && nRHTopo - nSeeds < threadsPerBlockForClustering) {
+          hcalFastCluster_multiSeedParallel(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 400 && nRHTopo - nSeeds <= 1500) {
+          hcalFastCluster_multiSeedIterative(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else {
+          if (once_per_block(acc))
+            printf("ERROR: Topo cluster %d has %d seeds and %d rechits. Will be processed in next kernel.\n",
+                   topoId,
+                   nSeeds,
+                   nRHTopo);
+        }
+      } else
+        return;
+    }
+  };
+
+  // Process very large, exotic topo clusters
+  class FastClusterExotic {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  const reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusterDeviceCollection::View clusterView,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView,
+                                  Position4* __restrict__ globalClusterPos,
+                                  Position4* __restrict__ globalPrevClusterPos,
+                                  float* __restrict__ globalClusterEnergy,
+                                  float* __restrict__ globalRhFracSum,
+                                  int* __restrict__ globalSeeds,
+                                  int* __restrict__ globalRechits) const {
+      const int nRH = pfRecHits.size();
+      int& topoId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& blockId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nRHTopo = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nSeeds = alpaka::declareSharedVar<int, __COUNTER__>(acc);
```

Comment:
These four can be local variables computed independently by each thread, instead of being stored in shared memory.